merlin-server/sql/merlin/tables/dataset.sql

```diff
@@ -32,3 +32,64 @@ create trigger delete_partitions_trigger
 exception
   when duplicate_object then null;
 end $$;
+
+create or replace function delete_dataset_cascade()
+  returns trigger
+  security definer
+  language plpgsql as $$begin
+    delete from span where span.dataset_id = old.id;
+    return old;
+end$$;
+
+create trigger delete_dataset_trigger
+  after delete on dataset
+  for each row
+  execute function delete_dataset_cascade();
+
+comment on trigger delete_dataset_trigger on dataset is e''
+  'Trigger to simulate an ON DELETE CASCADE foreign key constraint between span and dataset. The reason to'
+  'implement this as a trigger is that this single trigger can cascade deletes to any partitions of span.'
+  'If we used a foreign key, every new partition of span would need to add a new trigger to the dataset'
+  'table - which requires acquiring a lock that conflicts with concurrent inserts. In order to allow adding new'
+  'partitions concurrently with inserts to referenced tables, we have chosen to forego foreign keys from partitions'
+  'to other tables in favor of these hand-written triggers';
+
+create or replace function allocate_dataset_partitions(dataset_id integer)
+  returns dataset
+  security definer
+  language plpgsql as $$
+declare
+  dataset_ref dataset;
+begin
+  select * from dataset where id = dataset_id into dataset_ref;
+  if dataset_id is null
+  then
+    raise exception 'Cannot allocate partitions for non-existent dataset id %', dataset_id;
+  end if;
+
+  execute 'create table profile_segment_' || dataset_id || ' (
+    like profile_segment including defaults including constraints
+  );';
+  execute 'alter table profile_segment
+    attach partition profile_segment_' || dataset_id || ' for values in ('|| dataset_id ||');';
+
+  execute 'create table event_' || dataset_id || ' (
+      like event including defaults including constraints
+    );';
+  execute 'alter table event
+    attach partition event_' || dataset_id || ' for values in (' || dataset_id || ');';
+
+  execute 'create table span_' || dataset_id || ' (
+       like span including defaults including constraints
+    );';
+  execute 'alter table span
+    attach partition span_' || dataset_id || ' for values in (' || dataset_id || ');';
```

Comment:
Very minor: [The docs](https://www.postgresql.org/docs/15/plpgsql-statements.html#PLPGSQL-STATEMENTS-EXECUTING-DYN) recommend using variables and `USING` instead of concatenating the string, as "it avoids run-time overhead of converting the values to text and back". Was there a reason you chose to do concatenation instead?

Reply:
Thanks for sharing that - I was unaware of that option. I'll take a look at it this afternoon and see if I think it's a good fit

Reply:
I think the challenge here is that we're using the dataset_id to determine the name of the table itself - the docs I think are describing the ability to define variables that can be dynamically populated with values (similarly to how prepared statements work) - but I don't think that gives us the ability to use that variable in the name of the table

Reply:
I gave it a shot, but ran into this restriction mentioned in the docs:
> Another restriction on parameter symbols is that they only work in optimizable SQL commands (SELECT, INSERT, UPDATE, DELETE, MERGE, and certain commands containing one of these). In other statement types (generically called utility statements), you must insert values textually even if they are just data values.

Looks like we're out of luck for `CREATE TABLE` and `ALTER TABLE`

Reply:
That restriction is regarding the `$1` variables - format should still work

Reply:
Here's what it looks like with format - thoughts on whether the added verbosity helps readability? https://github.com/NASA-AMMOS/aerie/commit/52b7e641ec8de817810a1f27639708defc64da01?diff=split

Reply:
I don't think it does with how short out statements are. I do like the partition table names being created in advance, as it guarantees there's no typos and the hardest part to parse about the original was the concatenated table name, but it's not a really necessary change.

Reply:
Sounds good - I'll drop it from this PR - but we can keep it in mind if we find ourselves writing more complex `execute` statements. (The `execute` statements in an earlier iteration of this PR were much more involved - I'm glad they're not too bad now)